### PR TITLE
Update the implementation of DDM and Page-HinckleyDM

### PR DESCRIPTION
**Pull request by michaelchiucw**
Update the implementation of DDM and Page-HinckleyDM to follow the description in their paper respectively.

### CODE REVIEW

File: river/drift/ddm.py [hunk 1]
```diff
@@ -133,7 +133,7 @@ def update(self, value):
         self._in_warning_zone = False
         self.delay = 0
 
-        if self.sample_count < self.min_num_instances:
+        if self.sample_count <= self.min_num_instances:
```
Comment (jacobmontiel) [inline river/drift/ddm.py:136]: The current implementation is correct. DDM skips the drift detection until it has "seen" at least `min_num_instances`. This change is not necessary.

File: river/drift/page_hinkley.py [hunk 2]
```diff
@@ -86,12 +88,18 @@ def update(self, value) -> tuple:
 
         self.sample_count += 1
 
+        if self.sum <= self.Mint:
```
Comment (jacobmontiel) [inline river/drift/page_hinkley.py:91]: This is not necessary. We are already accounting for this in line 87. Based on the original paper:

image

**Review by jacobmontiel — CHANGES_REQUESTED** [review]
Thanks for the PR @michaelchiucw 

Nice catch on the test condition in DDM.

On the other hand, other changes are not necessary. Please check my comments.

File: river/drift/page_hinkley.py (hunk 2 above)
Comment (michaelchiucw) [reply river/drift/page_hinkley.py:91]: oh okay!! Thanks.

I seem to understand why they're equivalent now.
In eq.6, it defines `S'_0 = 0` and enforces the calculation on the positive side. So, the min(S_i) will always be 0 and we don't need to find the `self.Mint` anymore. Because min(S_i) is 0, so eq.5 (s_n - min(S_i) >= h) is equivalent to eq.7.

Do I missing or misunderstanding anything here? Many thanks!

File: river/drift/page_hinkley.py [hunk 3]
```diff
@@ -86,12 +88,18 @@ def update(self, value) -> tuple:
...
+
+        pht = self.sum - self.Mint
+
         self._in_concept_change = False
+        self._in_warning_zone = False
 
         if self.sample_count < self.min_instances:
             return False, False
 
-        if self.sum > self.threshold:
```
Comment (michaelchiucw) [inline river/drift/page_hinkley.py:94]: Also, here should be `if self.sum >= self.threshold:` according to the original paper. Am I right?

File: river/drift/ddm.py (hunk 1 above)
Comment (jacobmontiel) [reply river/drift/ddm.py:136]: Sorry for the delay. You are right. Although I expect the impact on the performance to be minimal, it is indeed different to the paper. Thanks for pointing this out.

File: river/drift/page_hinkley.py (hunk 2 above)
Comment (jacobmontiel) [reply river/drift/page_hinkley.py:91]: That is correct :-)

File: river/drift/page_hinkley.py (hunk 3 above)
Comment (jacobmontiel) [reply river/drift/page_hinkley.py:94]: Correct.

**Review by jacobmontiel — CHANGES_REQUESTED** [review]
Thanks for the explanation. Once the discussed changes are in place we can close the PR.

File: river/drift/page_hinkley.py (hunk 3 above)
Comment (michaelchiucw) [reply river/drift/page_hinkley.py:94]: I seem not able to the change the code from here. Should I close this pull request and make another one? thanks.

Comment (jacobmontiel) [reply river/drift/page_hinkley.py:94]: That is strange. Are you getting an error? Here is some useful information: 

https://github.com/online-ml/river/blob/master/CONTRIBUTING.md

You shouldn't need to create a new PR.

**Review by michaelchiucw — COMMENTED** [review]
I have just updated the Page-HinckleyDM.java according to the previous discussion.

File: river/drift/page_hinkley.py (hunk 3 above)
Comment (michaelchiucw) [reply river/drift/page_hinkley.py:94]: I have managed to find a way to edit it. Thanks!

**Comment by jacobmontiel** [comment]
I finally had time to look into why tests were failing. The test was triggering a corner case where a classifier is consistently wrong (`value=1`). Using the condition from the paper (`>=`) results in false drift signals every `num_min_instances`. I rolled back the condition and added a note for DDM regarding this change and the corner case. I also took the chance to update the test cases in the docstrings and the internal ones to generate more realistic data rather than the corner case mentioned.

I am closing the PR, sorry for the delay.

**Review by jacobmontiel — APPROVED** [review]
